File: server/controllers/ownerController.js
import asyncHandler from 'express-async-handler';
import { validationResult } from 'express-validator';
import PG from '../models/PG.js';
import User from '../models/User.js';
import cloudinary from '../config/cloudinary.js';
import { createNotification, emitToUser } from '../utils/notificationHelper.js';

// @desc   Get all PGs owned by logged in owner
// @route  GET /api/owner/pgs
export const getOwnerPGs = asyncHandler(async (req, res) => {
  const pgs = await PG.find({ owner: req.user._id })
    .sort({ createdAt: -1 });

  res.json({ success: true, count: pgs.length, data: pgs });
});

// @desc   Create new PG
// @route  POST /api/owner/pgs
export const createPG = asyncHandler(async (req, res) => {
  const errors = validationResult(req);
  if (!errors.isEmpty()) {
    return res.status(400).json({ success: false, message: 'Validation failed', errors: errors.array() });
  }

  const {
    title, description, price,
    city, area, address,
    roomType, genderPreference,
    amenities, contactNumber,
  } = req.body;

  // Collect uploaded image URLs from Cloudinary
  const images = req.files ? req.files.map((f) => f.path) : [];

  const pg = await PG.create({
    title,
    description,
    price: Number(price),
    location: { city, area, address },
    roomType,
    genderPreference,
    amenities: amenities ? (Array.isArray(amenities) ? amenities : amenities.split(',')) : [],
    images,
    contactNumber,
    owner: req.user._id,
    status: 'pending',
  });

  await createNotification(
    `New PG submitted: ${pg.title} by ${req.user.name}`,
    'ADD_PG',
    req.user._id,
    pg._id
  );

  res.status(201).json({ success: true, message: 'PG submitted for approval', data: pg });
});

// @desc   Update a PG
// @route  PUT /api/owner/pgs/:id
export const updatePG = asyncHandler(async (req, res) => {
  const pg = await PG.findById(req.params.id);

  if (!pg) {
    res.status(404);
    throw new Error('PG not found');
  }

  // Check ownership
  if (pg.owner.toString() !== req.user._id.toString() && req.user.role !== 'admin') {
    res.status(403);
    throw new Error('Not authorized to edit this PG');
  }

  if (pg.isDeleted) {
    res.status(404);
    throw new Error('PG has been deleted');
  }

  const {
    title, description, price,
    city, area, address,
    roomType, genderPreference,
    amenities, contactNumber,
    removeImages,
  } = req.body;

  // Handle image removal from Cloudinary
  if (removeImages) {
    const toRemove = Array.isArray(removeImages) ? removeImages : [removeImages];
    for (const imgUrl of toRemove) {
      // Extract public_id from URL
      const parts = imgUrl.split('/');
      const publicId = `sikkimpg/pgs/${parts[parts.length - 1].split('.')[0]}`;
      try {
        await cloudinary.uploader.destroy(publicId);
      } catch (err) {
        console.error('Cloudinary delete error:', err.message);
      }
    }
    pg.images = pg.images.filter((img) => !toRemove.includes(img));
  }

  // Add new uploaded images
  if (req.files && req.files.length > 0) {
    const newImages = req.files.map((f) => f.path);
    pg.images = [...pg.images, ...newImages].slice(0, 6);
  }

  pg.title = title || pg.title;
  pg.description = description || pg.description;
  pg.price = price ? Number(price) : pg.price;
  pg.location.city = city || pg.location.city;
  pg.location.area = area || pg.location.area;
  pg.location.address = address || pg.location.address;
  pg.roomType = roomType || pg.roomType;
  pg.genderPreference = genderPreference || pg.genderPreference;
  pg.amenities = amenities
    ? Array.isArray(amenities) ? amenities : amenities.split(',')
    : pg.amenities;
  pg.contactNumber = contactNumber || pg.contactNumber;
  pg.status = 'pending_update';

  await pg.save();

  await createNotification(
    `PG updated: "${pg.title}" needs re-approval`,
    'UPDATE_PG',
    req.user._id,
    pg._id
  );

  const savedByUsers = await User.find({ savedPGs: pg._id }).select('_id');
  await Promise.all(
    savedByUsers.map((u) =>
      createNotification(
        `A PG in your saved list was updated: "${pg.title}".`,
        'PG_STATUS_UPDATE',
        u._id,
        pg._id,
        { isAdminNotification: false }
      )
    )
  );

  res.json({ success: true, message: 'PG updated and pending re-approval', data: pg });
});

// @desc   Soft-delete a PG
// @route  DELETE /api/owner/pgs/:id
export const deletePG = asyncHandler(async (req, res) => {
  const pg = await PG.findById(req.params.id);

  if (!pg) {
    res.status(404);
    throw new Error('PG not found');
  }

  if (pg.owner.toString() !== req.user._id.toString() && req.user.role !== 'admin') {
    res.status(403);
    throw new Error('Not authorized to delete this PG');
  }

  pg.isDeleted = true;
  await pg.save();

  await createNotification(
    `PG deleted: "${pg.title}" by ${req.user.name}`,
    'DELETE_PG',
    req.user._id,
    pg._id
  );

  const savedByUsers = await User.find({ savedPGs: pg._id }).select('_id');
  await Promise.all(
    savedByUsers.map((u) =>
      createNotification(
        `A PG in your saved list was deleted: "${pg.title}".`,
        'PG_STATUS_UPDATE',
        u._id,
        pg._id,
        { isAdminNotification: false }
      )
    )
  );

  res.json({ success: true, message: 'PG deleted successfully' });
});

File: backend/app/controllers/owner_controller.py
from app.core.exceptions import AppError
from app.models import collections
from app.services.cloudinary_service import delete_images
from app.services.notification_service import create_notification
from app.utils.bson_utils import serialize_doc, serialize_many, to_object_id
from app.utils.datetime_utils import utc_now


async def get_owner_pgs_controller(db, current_user):
    cursor = db[collections.PGS].find({"owner": to_object_id(current_user["id"])}).sort("createdAt", -1)
    pgs = await cursor.to_list(length=1000)
    data = serialize_many(pgs)
    return {"success": True, "count": len(data), "data": data}


async def create_pg_controller(db, payload, current_user):
    now = utc_now()
    doc = {
        "title": payload.title,
        "description": payload.description,
        "price": payload.price,
        "location": {"city": payload.city, "area": payload.area, "address": payload.address},
        "roomType": payload.roomType.value,
        "genderPreference": payload.genderPreference.value,
        "amenities": payload.amenities,
        "rules": payload.rules,
        "images": payload.images,
        "contactNumber": payload.contactNumber,
        "owner": to_object_id(current_user["id"]),
        # Owner-approved users can list PGs directly.
        "status": "approved",
        "rejectionReason": "",
        "isDeleted": False,
        "averageRating": 0,
        "totalReviews": 0,
        "createdAt": now,
        "updatedAt": now,
    }
    result = await db[collections.PGS].insert_one(doc)
    doc["_id"] = result.inserted_id

    await create_notification(
        db,
        message=f"New PG listed by owner: {payload.title} by {current_user['name']}",
        notification_type="ADD_PG",
        user_id=to_object_id(current_user["id"]),
        related_id=result.inserted_id,
    )

    return {"success": True, "message": "PG listed successfully", "data": serialize_doc(doc)}


async def update_pg_controller(db, pg_id: str, payload, current_user):
    pg = await db[collections.PGS].find_one({"_id": to_object_id(pg_id)})
    if not pg:
        raise AppError(status_code=404, message="PG not found")
    if str(pg.get("owner")) != current_user["id"] and current_user.get("role") != "admin":
        raise AppError(status_code=403, message="Not authorized to edit this PG")
    if pg.get("isDeleted"):
        raise AppError(status_code=404, message="PG has been deleted")

    updates = {}
    if payload.title is not None:
        updates["title"] = payload.title
    if payload.description is not None:
        updates["description"] = payload.description
    if payload.price is not None:
        updates["price"] = payload.price
    if payload.roomType is not None:
        updates["roomType"] = payload.roomType.value
    if payload.genderPreference is not None:
        updates["genderPreference"] = payload.genderPreference.value
    if payload.amenities is not None:
        updates["amenities"] = payload.amenities
    if payload.rules is not None:
        updates["rules"] = payload.rules
    if payload.contactNumber is not None:
        updates["contactNumber"] = payload.contactNumber

    location = pg.get("location", {})
    if payload.city is not None:
        location["city"] = payload.city
    if payload.area is not None:
        location["area"] = payload.area
    if payload.address is not None:
        location["address"] = payload.address
    updates["location"] = location

    images = pg.get("images", [])
    if payload.removeImages:
        await delete_images(payload.removeImages)
        images = [img for img in images if img not in payload.removeImages]
    if payload.images:
        images = (images + payload.images)[:6]
    updates["images"] = images

    # Any edit must be approved by admin.
    updates["status"] = "pending_update"
    updates["updatedAt"] = utc_now()

    await db[collections.PGS].update_one({"_id": pg["_id"]}, {"$set": updates})
    updated = await db[collections.PGS].find_one({"_id": pg["_id"]})

    await create_notification(
        db,
        message=f"PG update requested: \"{updated['title']}\" needs admin approval",
        notification_type="UPDATE_PG",
        user_id=to_object_id(current_user["id"]),
        related_id=updated["_id"],
    )

    saved_by_users = await db[collections.USERS].find({"savedPGs": updated["_id"]}, {"_id": 1}).to_list(length=1000)
    for user in saved_by_users:
        await create_notification(
            db,
            message=f"A PG in your saved list was updated: \"{updated['title']}\".",
            notification_type="PG_STATUS_UPDATE",
            user_id=user["_id"],
            related_id=updated["_id"],
            is_admin_notification=False,
        )

    return {"success": True, "message": "PG update request sent to admin", "data": serialize_doc(updated)}


async def delete_pg_controller(db, pg_id: str, current_user, reason: str):
    pg = await db[collections.PGS].find_one({"_id": to_object_id(pg_id)})
    if not pg:
        raise AppError(status_code=404, message="PG not found")
    if str(pg.get("owner")) != current_user["id"] and current_user.get("role") != "admin":
        raise AppError(status_code=403, message="Not authorized to delete this PG")
    clean_reason = (reason or "").strip()
    if len(clean_reason) < 10:
        raise AppError(status_code=400, message="Please provide a delete reason with at least 10 characters")

    # Deletion must be approved by admin first.
    await db[collections.PGS].update_one(
        {"_id": pg["_id"]},
        {"$set": {"status": "pending_delete", "deleteRequestReason": clean_reason, "updatedAt": utc_now()}},
    )

    await create_notification(
        db,
        message=f"PG delete requested: \"{pg['title']}\" by {current_user['name']}. Reason: {clean_reason}",
        notification_type="DELETE_PG",
        user_id=to_object_id(current_user["id"]),
        related_id=pg["_id"],
    )

    return {"success": True, "message": "PG deletion request sent to admin"}
